### PR TITLE
[XLA] Unimplemented inst in call fusion

### PR DESCRIPTION
**Pull request by DavidNorman**
If an unimplemented instruction is embedded inside a call or fusion operation then currently the system aborts having done a ConsumeValueOrDie.  This change allows the various clients of the evaluator to process a graph containing calls/fusions safely.

Applies to all XLA systems.

### CODE REVIEW

File: tensorflow/compiler/xla/service/hlo_evaluator_test.cc [hunk 1]
```diff
@@ -2924,5 +2924,59 @@ ENTRY main {
             "Expected 1 argument, but got 2.");
 }
 
+// Check that we correctly return error status when evaluating call instruction
+// containing an unimplemented instruction
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/hlo_evaluator_test.cc]: Nit

> Check that we correctly return *an* error status when evaluating *a* call instruction containing an unimplemented instruction.

Nit: Period at end of sentence

File: tensorflow/compiler/xla/service/hlo_evaluator_test.cc [hunk 2]
```diff
@@ -2924,5 +2924,59 @@ ENTRY main {
...
+  TF_ASSERT_OK_AND_ASSIGN(m_, ParseAndReturnVerifiedModule(hlo_text));
+
+  auto* comp = m_->entry_computation();
+  auto* call = comp->GetInstructionWithName("r");
+
+  Literal lit;
+  EXPECT_FALSE(HloEvaluator().TryEvaluate(call, &lit));
+}
+
+// Check that we correctly return error status when evaluating call instruction
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/hlo_evaluator_test.cc]: s/call/fusion/

**Review by jlebar — CHANGES_REQUESTED** [review]
Thank you for the patch!

As it happens, these testcases are about to break because we're about to land a patch to make kRNG work in the evaluator.

Is there another instruction you could use?

**Comment by DavidNorman** [comment]
used a custom-call

**Comment by jlebar** [comment]
> is there a danger that the constant subexpression evaluator will eval the kRng instructions and eliminate them from the graph?

CSE doesn't evaluate instructions, it just commons them (i.e. replaces two instances of the same thing with one instance).  So the evaluator isn't used in CSE, and however it worked before it should continue working.

CSE skips side-effecting ops, and kRNG is marked as side-effecting.  As it happens @bixia1 is working on changing this and making the side-effects more explicit in the HLO graph, e.g. passing the state (i.e. the seed) explicitly between kRNG ops.

**Review by jlebar — APPROVED** [review]

**Comment by DavidNorman** [comment]
@jlebar sorry - i meant the constant subexpression removal step (which uses the evaluator) rather than the common subexpression removal.

**Comment by jlebar** [comment]
> sorry - i meant the constant subexpression removal step (which uses the evaluator) rather than the common subexpression removal.

Ah, what we call "constant folding".  `hlo_constant_folding.cc` explicitly skips kRNG ops.  Which I suppose is a bit of a hack, but there you go.

**Comment by ymodak** [comment]
@DavidNorman Can you please resolve the merge conflicts? Thanks!

**Comment by ymodak** [comment]
@DavidNorman gentle ping

**Pull request closed without merging**